cmd/skaffold/app/cmd/flags.go

```diff
@@ -0,0 +1,111 @@
+	commonFlags.StringVarP(&opts.DefaultRepo, "default-repo", "d", "", "Default repository value (overrides global config)")
+	return commonFlags
+}
+
+func buildFlagSet(name string) *flag.FlagSet {
+	buildFlags := flag.NewFlagSet(name, flag.ContinueOnError)
+	buildFlags.BoolVar(&opts.CacheArtifacts, "cache-artifacts", false, "Set to true to enable caching of artifacts.")
+	buildFlags.StringVarP(&opts.CacheFile, "cache-file", "", "", "Specify the location of the cache file (default $HOME/.skaffold/cache)")
+	buildFlags.StringArrayVar(&opts.InsecureRegistries, "insecure-registry", nil, "Target registries for built images which are not secure")
+	return buildFlags
+}
+
+func eventsAPIFlagSet(name string) *flag.FlagSet {
+	eventFlags := flag.NewFlagSet(name, flag.ContinueOnError)
+	eventFlags.BoolVar(&opts.EnableRPC, "enable-rpc", false, "Enable gRPC for exposing Skaffold events (true by default for `skaffold dev`)")
+	eventFlags.IntVar(&opts.RPCPort, "rpc-port", constants.DefaultRPCPort, "tcp port to expose event API")
+	eventFlags.IntVar(&opts.RPCHTTPPort, "rpc-http-port", constants.DefaultRPCHTTPPort, "tcp port to expose event REST API over HTTP")
+	return eventFlags
+}
+
+func deployFlagSet(name string) *flag.FlagSet {
+	deployFlags := flag.NewFlagSet(name, flag.ContinueOnError)
+	deployFlags.BoolVar(&opts.Tail, "tail", false, "Stream logs from deployed objects")
+	deployFlags.BoolVar(&opts.Force, "force", false, "Recreate kubernetes resources if necessary for deployment (default: false, warning: might cause downtime!)")
+	deployFlags.StringArrayVarP(&opts.CustomLabels, "label", "l", nil, "Add custom labels to deployed objects. Set multiple times for multiple labels.")
+	deployFlags.BoolVar(&opts.Notification, "toot", false, "Emit a terminal beep after the deploy is complete")
+	return deployFlags
+}
+
+func testFlagSet(name string) *flag.FlagSet {
+	testFlags := flag.NewFlagSet(name, flag.ContinueOnError)
+	testFlags.BoolVar(&opts.SkipTests, "skip-tests", false, "Whether to skip the tests after building")
+	return testFlags
+}
+
+func cleanupFlagSet(name string) *flag.FlagSet {
+	cleanupFlags := flag.NewFlagSet(name, flag.ContinueOnError)
+	cleanupFlags.BoolVar(&opts.Cleanup, "cleanup", true, "Delete deployments after dev or debug mode is interrupted")
+	cleanupFlags.BoolVar(&opts.NoPrune, "no-prune", false, "Skip removing images and containers built by Skaffold")
+	return cleanupFlags
+}
+
+func AddFlags(cmd *cobra.Command) {
+	cmd.Flags().AddFlagSet(CommonFlagSet)
+	cmd.Flags().AddFlagSet(getAnnotatedFlags(cmd.Use, cmd.Annotations))
+}
+
+func getAnnotatedFlags(name string, annotations map[string]string) *flag.FlagSet {
+	allFlags := flag.NewFlagSet(name, flag.ContinueOnError)
+	for a := range annotations {
+		flags, ok := AnnotationToFlag[a]
+		if ok {
+			allFlags.AddFlagSet(flags)
+		}
+	}
+	return allFlags
+}
+
+func addTailDevFlag(cmd *cobra.Command) {
```

Comment:
Why is this a separate function, instead of an annotation, like `devFlagSet`?

Reply:
This is  because of existing flag "--tail" is used for setting "opts.TailDev" and "opts.Tail".
In Deploy flow, "--tail" is set to opts.Tail with default set to false
However, in run and dev flow, "--tail" is set to opts.TailDev with default set to true

Because the code adds annotations as map, the order of keys 
```
		Annotations: map[string]string{
			BuildAnnotation:   "true",
			DeployAnnotation:  "true",
			TestAnnotation:    "true",
			CleanupAnnotation: "true",
			EventsAnnotation:  "true",
		},
	}
```
Dev and Build, also run deploy, so it gets messy. 
So inorder to solve this, in Dev and Run flow, we first register the "--tail" flag and bind it to opts.TaildDev.
Then in the `AddFlags` function when we register the same flag, it ignores it.

Reply:
Ideally, this should have been another flag like "--tail-dev" or a way to set defaults for "--tail" flag based on the command. But since flagSets do not have command reference, there isn't a way to do it.